src/kafka-util/src/aws.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2024 Yuhao Su. All rights reserved.
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// This file is derived from:
+//
+//    https://github.com/yuhao-su/aws-msk-iam-sasl-signer-rs
+//
+// It was incorporated directly into Materialize on August 1, 2024.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! AWS integration for Kafka.
+
+use std::time::{Duration, SystemTime};
+
+use aws_credential_types::provider::error::CredentialsError;
+use aws_credential_types::{provider::ProvideCredentials, Credentials};
+use aws_types::{region::Region, sdk_config::SharedCredentialsProvider};
+use base64::prelude::BASE64_URL_SAFE_NO_PAD;
+use base64::Engine;
+use chrono::NaiveDateTime;
+use thiserror::Error;
+use url::Url;
+
+/// The default expiration time in seconds.
+const DEFAULT_EXPIRY_SECONDS: u32 = 900;
+
+/// An error while signing an AWS IAM URL.
+#[derive(Error, Debug)]
+pub enum SignerError {
+    /// An error while fetching AWS credentials.
+    #[error("failed to provide credentials: {0}")]
+    ProvideCredentials(#[from] CredentialsError),
+    /// An error constructing the authentication token.
+    #[error("failed constuct auth token: {0}")]
+    ConstructAuthToken(String),
+}
+
+/// Generate a base64-encoded signed url as an auth token by loading IAM
+/// credentials from an AWS credentials provider.
+pub async fn generate_auth_token(
+    region: Region,
+    credentials_provider: SharedCredentialsProvider,
+) -> Result<(String, i64), SignerError> {
+    let credentials = credentials_provider.provide_credentials().await?;
+    construct_auth_token(region, credentials)
+}
+
+fn construct_auth_token(
+    region: Region,
+    credentials: Credentials,
+) -> Result<(String, i64), SignerError> {
+    let endpoint_url = format!("https://kafka.{}.amazonaws.com", region);
```

Comment:
shouldn't this be configured based on the endpoint_url of the AWS Connection object? I don't know if any other providers follow the MSK API but we do allow customers to set the endpoint here https://materialize.com/docs/sql/create-connection/#aws-options and it seems unusual to ignore it for this auth check

Reply:
There's also some non standard endpoints in AWS for FIPs I think predominantly used in gov cloud. I don't think we need to seriously consider supporting these, but it's worth calling out the limitations of this endpoint construction. 
https://docs.aws.amazon.com/general/latest/gr/rande.html#FIPS-endpoints

Reply:
Good point! I do think in theory we want to pull this from the AWS configuration in the connection, but the AWS SDK is inscrutable on how to do this. I dropped a comment in but I'm just going to leave this as is for now. AFAIK nothing besides real AWS supports MSK IAM auth, so there's no way to test this with localstack or anything ([localstack ticket](https://github.com/localstack/localstack/issues/10617)), and so the hardcoding here is ~fine. (Yeah, doesn't support FIPS, but that's a problem for a long time for now.)